### PR TITLE
You Only Query Once

### PR DESCRIPTION
**Pull request by DanielNilssonAISweden**
# Description

Summary of changes

* Added You Only Query Once (YOQO) attack.

### CODE REVIEW

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        true labels, and signal values.
+
+        """
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size=1)
+
+        predictions = []
+        signal_values = []
+
+        device_name = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.mps.is_available() else "cpu")
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:300]: Not all torch versions have `mps.is_available()`. Perhaps check `torch.has_mps `and if true, look for mps.

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            if self.output_shape == 1:
+                self._optimization_objective = self._optimization_objective_offline_binary
+            else:
+                self._optimization_objective = self._optimization_objective_offline
+
+        self.weights = 1 + (self.alpha - 1) * self.in_indices_masks
+        self.weights = Tensor(self.weights)
+        self.batch_size = 5000
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size=self.batch_size)
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:196]: Please add comments for the code below.

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+                labels_false[:,np.arange(labels_true.shape[0]),labels_true] = -1e10
+                labels_false = torch.argmax(labels_false, dim = 2)
+                batch_target_output = \
+                    labels_true[:,np.newaxis] * self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:] + \
+                    labels_false.T * (1 - self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:])
+                self.target_output.extend(batch_target_output)
+        self.target_output = np.array(self.target_output)
+        self.target_output = Tensor(self.target_output)
+
+    def _optimization_objective_online(
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:216]: could be written as a single function with extra argument

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        loss = self.loss(loss, target_output)
+        loss = loss * weights
+        return loss.sum()
+
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:239]: same as above.

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+            self.mse = MSELoss()
+
+            if self.output_shape == 1:
+                self._optimization_objective = self._optimization_objective_offline_binary
+            else:
+                self._optimization_objective = self._optimization_objective_offline
+
+        self.weights = 1 + (self.alpha - 1) * self.in_indices_masks
+        self.weights = Tensor(self.weights)
+        self.batch_size = 5000
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:194]: This may result in a very large memory allocation. Perhaps stick with default value in `get_dataloader()`?

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+                self._optimization_objective = self._optimization_objective_online
+        else:
+            self.mse = MSELoss()
+
+            if self.output_shape == 1:
+                self._optimization_objective = self._optimization_objective_offline_binary
+            else:
+                self._optimization_objective = self._optimization_objective_offline
+
+        self.weights = 1 + (self.alpha - 1) * self.in_indices_masks
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:192]: add comment: weight for out-models : alpha, weight for in-models: 1

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        self.weights = Tensor(self.weights)
+        self.batch_size = 5000
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size=self.batch_size)
+        self.target_output = []
+        for i, (data, labels) in tqdm(enumerate(data_loader), desc="Calculating target labels", leave=False):
+            if self.output_shape == 1:
+                batch_target_output = \
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:200]: add comment: associate true labels with in-models and false labels with out-models

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+                    labels[:,np.newaxis] * self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:] + \
+                    (1 - labels[:,np.newaxis]) * (1 - self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:])
+                self.target_output.extend(batch_target_output)
+            else:
+                labels_true = labels
+                labels_false = torch.stack([shadow_model.model_obj(data) for shadow_model in self.shadow_models])
+                labels_false[:,np.arange(labels_true.shape[0]),labels_true] = -1e10
+                labels_false = torch.argmax(labels_false, dim = 2)
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:208]: comment: associate erroneous label with largest logit score to out-models

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        predictions = []
+        signal_values = []
+
+        device_name = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.mps.is_available() else "cpu")
+
+        for model in self.shadow_models:
+            model.model_obj.eval()
+            model.model_obj.to(device_name)
+
+        for i, (data, labels) in tqdm(enumerate(data_loader), desc="Optimizing queries", leave=False):
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:306]: show progress bar? Right now the user does not know how much time remains.

File: leakpro/attacks/mia_attacks/yoqo.py (hunk 6 above)
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:192]: This assigns alpha to in-models (self.in_indices_masks = True if a sample is in the training set of the shadow model). Hence, the out-models will get weight=1. 

This makes the offline attack seemless as the alpha is simply added to the MSE but the online attack does not align with the eq. 4. Perhaps make a comment that the in-models are weighted rather than the out-models. This is just to ensure no confusion if one attempts to play around with alpha.

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+            lprime = np.argmax(lprime, axis = 1)
+
+            batch_predictions = np.equal(lprime,labels)
+            predictions.extend(batch_predictions)
+
+        predictions = np.array(predictions).reshape(1,-1)
+        signal_values = predictions.copy().reshape(-1,1)
+
+        for model in self.shadow_models:
+            model.model_obj.train()
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:321]: Is this needed?

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+
+        # Prepare true labels array, marking 1 for training data and 0 for non-training data
+        true_labels = np.concatenate(
+            [np.ones(len(self.in_members)), np.zeros(len(self.out_members))]
+        )
+
+        logger.info(f"Accuracy: {np.sum(predictions == true_labels)/predictions.size}")
+
+        predictions = np.concatenate(
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:330]: why padding zeros and ones?

**Review by johanos1 — CHANGES_REQUESTED** [review]
Nice job! I like how you manage to handle the offline and online simultaneously. I have a few comments before ready for merge. Let me know if anything is unclear.

File: leakpro/attacks/mia_attacks/yoqo.py (hunk 1 above)
Comment (DanielNilssonAISweden) [reply leakpro/attacks/mia_attacks/yoqo.py:300]: It's perhaps cleaner to just remove the check? Since there are no checks for MPS in the rest of the code anyway.

File: leakpro/attacks/mia_attacks/yoqo.py (hunk 6 above)
Comment (DanielNilssonAISweden) [reply leakpro/attacks/mia_attacks/yoqo.py:192]: Oh, this is just a mistake on my part, I had not realized that I had switched alpha. Will fix so the online method matches the paper.

File: leakpro/attacks/mia_attacks/yoqo.py (hunk 11 above)
Comment (DanielNilssonAISweden) [reply leakpro/attacks/mia_attacks/yoqo.py:330]: This is needed for the report generation. Otherwise I think there was a crash on trying to generate the ROC curve. Will add a comment.

File: leakpro/attacks/mia_attacks/yoqo.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        if self.binary_output:
+            loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        else:
+            loss = torch.stack([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 2)
+        loss = self.loss(loss, target_output)
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/yoqo.py:225]: For Binary cross entropy, this crashes as target_output must be float.
For Crossentropy, the target should be long so it works.

**Review by johanos1 — CHANGES_REQUESTED** [review]
